ob1k-consul/src/main/java/com/outbrain/ob1k/consul/CooperativeTargetProvider.java

```diff
@@ -0,0 +1,126 @@
+    // may serve for double dispatch. That's why we increment the round robin counter
+    // by one instead of targetsNum.
+    final int index = roundRobin.getAndIncrement();
+
+    // Maintain the order of the elements so that we can tie break them
+    // by the round robin order.
+    final Set<String> providedTargetsSet = new LinkedHashSet<>();
+
+    // limit the target pool size to keep the complexity O(targetsNum).
+    final int targetsPool = Math.min(targetsNum + EXTRA_TARGETS, currTargets.size());
+    int maxPendingRequests = 0;
+    for (int i = 0; i < targetsPool; i++) {
+      final String target = currTargets.get(Math.abs((index + i) % currTargets.size()));
+      if (i == 0) {
+        decreasePenalty(target);
+      }
+      providedTargetsSet.add(target);
+      maxPendingRequests = Integer.max(maxPendingRequests, pendingRequests.count(target));
+    }
+    this.maxPendingRequests = maxPendingRequests;
+    return providedTargetsSet;
+  }
+
+  private void decreasePenalty(String target) {
+    final int remainingPenalty = this.remainingPenalty.count(target);
+    if (remainingPenalty > 0) {
+      // Using the least of remainingPenalty - 1, penalize(target) is done to handle the case
+      // where a target was penalized heavily because the max number of
+      // pending requests was high at the time, but then the maximum decreased
+      // markedly, and its penalty is now too high.
+      this.remainingPenalty.setCount(target, Integer.min(remainingPenalty - 1, penalize(target)));
+    }
+  }
+
+  private List<String> bestTargets(int targetsNum, Set<String> providedTargetsSet) {
+    // create snapshots so that the comparator will not switch the order of targets during sorting.
+    Map<String, Integer> pendingRequests = Maps.newHashMapWithExpectedSize(providedTargetsSet.size());
+    providedTargetsSet.forEach(target -> pendingRequests.put(target, this.pendingRequests.count(target)));
+
+    Map<String, Integer> remainingPenalty = Maps.newHashMapWithExpectedSize(providedTargetsSet.size());
+    providedTargetsSet.forEach(target -> remainingPenalty.put(target, this.remainingPenalty.count(target)));
+
+    Map<String, Integer> roundRobinPosition = Maps.newHashMapWithExpectedSize(providedTargetsSet.size());
+    providedTargetsSet.forEach(target -> roundRobinPosition.put(target, roundRobinPosition.size()));
+
+    return Ordering.from(Comparator.<String, Integer>comparing(target -> remainingPenalty.get(target) + pendingRequests.get(target)).
+                    thenComparing(roundRobinPosition::get)).
+            leastOf(providedTargetsSet, targetsNum);
+  }
+
+  @Override
+  public void onTargetsChanged(List<HealthInfoInstance> healthTargets) {
+    super.onTargetsChanged(healthTargets);
+
+    // Clear remainingPenalty, since we may never encounter
+    // any of these targets again.
+    remainingPenalty.clear();
+  }
+
+  public void targetDispatched(String target) {
```

Comment:
This will require the calling code to be aware of the implementation, understand the internals here, and cooperate... 
This isn't part of the interface